src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,502 @@
+  Contains the `nil` introduced for an `ident` by the `state-after-top-level-ident-dissoc`"
+  [state-map ident]
+  (let [value-at-path (fn [a-path]
+                        (if (>= (count a-path) 4)
+                          ;; don't follow idents for denormalized paths
+                          (clojure.core/get-in (state-after-top-level-ident-dissoc state-map ident)
+                                               a-path)
+                          ;; follow idents for denormalized paths
+                          (get-in (state-after-top-level-ident-dissoc state-map ident)
+                                  a-path)))]
+    (map (fn [a-path]
+           (if (map? (value-at-path a-path))
+             ;; finds db-path from the original app-db
+             (tree-path->db-path state-map a-path)
+             (value-at-path a-path)))
+         (all-paths-after-top-level-dissoc state-map ident))))
+
+
+(defn- entity-path-value-map-after-top-level-dissoc
+  "Returns a map of all path vectors and their corresponding values.
+  Contains the paths and their corresponding values (both normalized and denormalized).
+  The values contain the `nil` introduced for an `ident` by the `state-after-top-level-ident-dissoc`"
+  [state-map ident]
+  (zipmap (all-paths-after-top-level-dissoc state-map ident)
+          (all-values-at-path-after-top-level-dissoc state-map ident)))
+
+
+(defn- prune-ident
+  "This is the reducing function used to prune the `nil` values which are the dangling pointers to
+  a an entity which is already removed."
+  [state-map [a-path a-value] ident]
+  ;; if denormalized path, do nothing
+  (if (>= (count a-path) 4)
+    state-map
+    (cond
+      (nil? a-value) (dissoc-in state-map a-path)
+
+      (vector-of-vectors? a-value) (assoc-in state-map a-path
+                                             (apply vector (remove #{ident} a-value)))
+      :else state-map)))
+
+;;====
+
+(>defn remove-entity*
+  "Remove the given entity at the given ident. Also scans all tables and removes any to-one or to-many idents that are
+   found that match `ident` (removes dangling pointers to the removed entity).
+
+   The optional `cascade` parameter is a set of keywords that represent edges that should cause recursive deletes
+   (i.e. it indicates edge names that *own* something, indicating it is safe to remove those entities as well).
+
+   Returns the new state map with the entity(ies) removed."
+
+  ([state-map ident]
+   [map? eql/ident? => map?]
+   (remove-entity* state-map ident #{}))
+
+  ;;TODO implement the cascading feature
+
+  ([state-map ident cascade]
+   [map? eql/ident? (s/coll-of keyword? :kind set?) => map?]
+
+   (reduce #(prune-ident %1 %2 ident)
+           (state-after-top-level-ident-dissoc state-map ident)
+           (entity-path-value-map-after-top-level-dissoc state-map ident))))
+
+
+(comment
+
+  (def state {:fastest-car  [:car/id 1]
```

Comment:
OK, so on the plus side I understand that you want to be able to "follow" the data. Making tests comprehensible is super great. Love it.

But, this is a rather large a data structure all clumped together. Each behavior of the function only needs a small portion of this data to be "proven", so I'd suggest, if possible, breaking this up into small structures that are focused to the behavior under test, and co-locate the small thing near the small test.  Perhaps I'm wrong on this, but it's my first impression (whoa! That's a lot to comprehend to figure out a single failing test)

Reply:
Yup, just needed to be on the same page with you regarding the entities to be removed. Now that we're here, I'll shrink these for sure 😊